OFED builder workflow

assumptionsandg opened the pull request.

assumptionsandg committed:
OFED workflow

markgoddard commented on .github/workflows/package-build-ofed.yml:
```diff
@@ -0,0 +1,197 @@
+        with:
+          path: src/kayobe-config
+
+      - name: Output image tag of the builder
+        id: builder_image_tag
+        run: |
+          echo image_tag=$(grep stackhpc_rocky_9_overcloud_host_image_version: etc/kayobe/pulp-host-image-versions.yml | awk '{print $2}') >> $GITHUB_OUTPUT
+
```
Task not required.

holliefae replied:
Removed

markgoddard commented on .github/workflows/package-build-ofed.yml:
```diff
@@ -0,0 +1,197 @@
+          source venvs/kayobe/bin/activate &&
+          source src/kayobe-config/kayobe-env --environment ci-builder &&
+          kayobe seed host configure -e seed_bootstrap_user=rocky --skip-tags network
+
```
Do we need to reboot at this point to boot into the release train kernel? Or is having the packages available enough to build against?

markgoddard replied:
Design summit notes say:

> Ensure target kernel is running (it has to be the only kernel on the VM)

holliefae replied:
Running on the latest kernel from the configured release train repo now.

Alex-Welsh replied:
Reboot is applied during build-ofed.yml playbook

holliefae replied:
Moved reboot to the workflow in a941098b

markgoddard commented on etc/kayobe/ansible/build-ofed.yml:
I'd suggest separating the build and upload steps into different playbooks.

holliefae replied:
Pulp uploads are now in push-ofed.yml

markgoddard commented on etc/kayobe/ansible/build-ofed.yml:
```diff
@@ -0,0 +1,111 @@
+    - name: Download MellanoxOFED archive
+      ansible.builtin.get_url:
+        url: https://content.mellanox.com/ofed/MLNX_OFED-{{ stackhpc_pulp_mlnx_ofed_version }}/{{ stackhpc_mlnx_ofed_file_string }}.tgz
+        dest: /opt/ofed/ofed-archive
```
.tgz?

holliefae replied:
Seems to be the correct URL

Alex-Welsh replied:
Should the dest have a .tgz suffix?

holliefae replied:
Added, it was working without it though. Does look more normal this way

Alex-Welsh replied:
Yeah, it's just nice to keep the extension for clarity

markgoddard commented on etc/kayobe/ansible/build-ofed.yml:
```diff
@@ -0,0 +1,111 @@
+        password: "{{ stackhpc_release_pulp_password }}"
+        name: "{{ stackhpc_pulp_repo_mlnx_ofed.name }}"
+        state: present
+      retries: "{{ pulp_repository_rpm_repositories_retries }}"
```
Is pulp_repository_rpm_repositories_retries defined here?

holliefae replied:
Defaults to 3, Not sure if we should just hardcode that or leave pulp_timeout_retries to be defined when required.

markgoddard commented on etc/kayobe/pulp.yml:
```diff
@@ -369,6 +369,16 @@ stackhpc_pulp_repository_rpm_repos: >-
   {%- endfor -%}
   {{ rpm_repos }}
 
+# Mellanox OFED version
+stackhpc_pulp_mlnx_ofed_version: 24.04-0.7.0.0
+
+# Mellanox OFED repositories
+stackhpc_pulp_repo_mlnx_ofed:
+  name: Mellanox Technologies mlnx_ofed {{ stackhpc_pulp_mlnx_ofed_version }}
+  url: "{{ stackhpc_release_pulp_content_url }}/mlnx_ofed/{{ stackhpc_pulp_mlnx_ofed_version }}/rhel9.{{ stackhpc_pulp_repo_rocky_9_minor_version }}/x86_64/{{ stackhpc_pulp_repo_mlnx_ofed_version }}"
+  distribution_name: "mlnx_ofed_{{ stackhpc_pulp_mlnx_ofed_version }}-"
+  base_path: "mlnx_ofed/{{ stackhpc_pulp_mlnx_ofed_version }}/rhel9.{{ stackhpc_pulp_repo_rocky_9_minor_version }}/x86_64/"
+
```
This should move to a separate section, just before containers on L426.

holliefae replied:
Moved to MellanoxOFED section

markgoddard commented on etc/kayobe/pulp.yml:
```diff
@@ -369,6 +369,16 @@ stackhpc_pulp_repository_rpm_repos: >-
+stackhpc_pulp_mlnx_ofed_version: 24.04-0.7.0.0
+
+# Mellanox OFED repositories
+stackhpc_pulp_repo_mlnx_ofed:
```
Needs rhel9 in the variable name?

Alex-Welsh replied:
Done

markgoddard commented on etc/kayobe/pulp.yml:
```diff
@@ -369,6 +369,16 @@ stackhpc_pulp_repository_rpm_repos: >-
+
+# Mellanox OFED repositories
+stackhpc_pulp_repo_mlnx_ofed:
+  name: Mellanox Technologies mlnx_ofed {{ stackhpc_pulp_mlnx_ofed_version }}
```
Needs to include the Rocky minor version in the name:

> Use one Pulp RPM repo for each version of OFED and Rocky minor version

holliefae replied:
Added Rocky9.x to repo name.

Alex-Welsh replied:
Has this been reverted @assumptionsandg still looks the same to me

markgoddard commented on etc/kayobe/pulp.yml:
```diff
@@ -369,6 +369,16 @@ stackhpc_pulp_repository_rpm_repos: >-
+stackhpc_pulp_repo_mlnx_ofed:
+  name: Mellanox Technologies mlnx_ofed {{ stackhpc_pulp_mlnx_ofed_version }}
+  url: "{{ stackhpc_release_pulp_content_url }}/mlnx_ofed/{{ stackhpc_pulp_mlnx_ofed_version }}/rhel9.{{ stackhpc_pulp_repo_rocky_9_minor_version }}/x86_64/{{ stackhpc_pulp_repo_mlnx_ofed_version }}"
+  distribution_name: "mlnx_ofed_{{ stackhpc_pulp_mlnx_ofed_version }}-"
```
Needs to include the Rocky minor version in the name:

markgoddard replied:
nit: distribution names normally uses dashes.

holliefae replied:
Done.

markgoddard commented on .github/workflows/package-build-ofed.yml:
```diff
@@ -0,0 +1,197 @@
+          ssh_public_key = "id_rsa.pub"
+          ssh_username = "rocky"
+          aio_vm_name = "skc-host-image-builder"
+          # Must be a Rocky Linux 9 host to successfully build all images
+          # This MUST NOT be an LVM image. It can cause confusing conficts with the built image.
```
nit: comment does not apply to OFED builds AFAIK

Alex-Welsh replied:
Comment has been removed

markgoddard reviewed (changes requested):
Thanks @assumptionsandg this is a nice start.

Alex-Welsh commented:
@assumptionsandg is this tested & ready for review?

assumptionsandg commented:
> @assumptionsandg is this tested & ready for review?

The playbooks are tested, yes. The workflow is not.

assumptionsandg committed:
Support DOCA OFED

assumptionsandg committed:
Push kernel modules

assumptionsandg committed:
Push OFED userspace packages

assumptionsandg committed:
Fix build

assumptionsandg committed:
Fix kernel upgrade

assumptionsandg committed:
Replace MLNX with DOCA

assumptionsandg committed:
Adjust lv_var_tmp to 2G

assumptionsandg committed:
Fix workflow

assumptionsandg committed:
Disable gpg check for doca host

assumptionsandg committed:
Fix inputs in workflow

assumptionsandg committed:
Replace with_fileglob

assumptionsandg committed:
Remove trailing slash in base_path

assumptionsandg committed:
Install kernel modules

assumptionsandg committed:
Re-add the vault password

assumptionsandg committed:
Remove trailing dash from distribution

Alex-Welsh commented on etc/kayobe/ansible/build-ofed.yml:
```diff
@@ -0,0 +1,103 @@
+        state: latest
+        update_cache: true
+
+    - name: Set GRUB default index to saved
```
Could we use our [grub command line](https://github.com/stackhpc/ansible-collection-linux/tree/main/roles/grubcmdline) collection for this?

holliefae replied:
I think that role is for updating the kernel command line, but here we're setting the default kernel entry for GRUB to boot.

Alex-Welsh commented on etc/kayobe/ansible/build-ofed.yml:
```diff
@@ -0,0 +1,103 @@
+        line: "UPDATEDEFAULT=yes"
+        create: true
+
+    - name: Update the default kernel entry
```
would the reset-bls-entries.yml playbook help here?

holliefae replied:
We can probably use this, though that playbook runs on overcloud and not the seed.

We'll also need to split build-ofed up, but that's probably a good thing as the kernel update/ofed build should probably be separate.

holliefae replied:
This is implemented in a941098b, let me know what you think.

Alex-Welsh commented on etc/kayobe/ansible/build-ofed.yml:
```diff
@@ -0,0 +1,103 @@
+      ansible.builtin.reboot:
+        reboot_timeout: 600
+
+    - name: Add DOCA host repository
```
The name of this task doesn't appear to match the content. Is it installing a package that includes the repo?

holliefae replied:
Yes, that package installs the doca-host repository which is used for the doca-ofed installation process. The name could be clearer though.

holliefae replied:
Changed to `Add DOCA host repository package` in a941098b.

Alex-Welsh commented on etc/kayobe/ansible/build-ofed.yml:
```diff
@@ -0,0 +1,103 @@
+    - name: Add DOCA host repository
+      ansible.builtin.dnf:
+        name: https://developer.nvidia.com/downloads/networking/secure/doca-sdk/DOCA_2.8/doca-host-2.8.0-204000_{{ stackhpc_pulp_doca_ofed_version }}_rhel9{{ stackhpc_pulp_repo_rocky_9_minor_version }}.x86_64.rpm
+        disable_gpg_check: true
```
This is a bit sketchy. It should be fine since it's an `https` url but still would be better to enable the check

holliefae replied:
I agree, however the package is not signed, so will fail on install.

Alex-Welsh commented on etc/kayobe/ansible/build-ofed.yml:
```diff
@@ -0,0 +1,103 @@
+    - name: Set build directory
+      ansible.builtin.lineinfile:
+        path: /opt/mellanox/doca/tools/doca-kernel-support
+        search_string: 'TMP_DIR=$1'
```
If TMP_DIR is $1 can we not just set $1 to `/home/cloud-user/ofed`?

holliefae replied:
This is an argument to the `compile_install_ofed` function in the doca-kernel-support bash script and not the script as a whole. This script only provides a kernel argument and has no way to configure the build directory.

Alex-Welsh commented on etc/kayobe/ansible/build-ofed.yml:
```diff
@@ -0,0 +1,103 @@
+      ansible.builtin.lineinfile:
+        path: /opt/mellanox/doca/tools/doca-kernel-support
+        search_string: 'TMP_DIR=$1'
+        line: '    TMP_DIR=/home/cloud-user/ofed'
```
Is that whitespace required?
```suggestion
        line: 'TMP_DIR=/home/cloud-user/ofed'
```

holliefae replied:
For indentation yes, though maybe we can use replace instead of lineinfile.

holliefae replied:
Replaced with replace in 2be44f99

assumptionsandg committed:
Remove LVM configuration

LVM config is already applied for the seed in the ci-builder
environment, and this should be enough for building OFED.

assumptionsandg committed:
Use reset-bls-entries playbook in OFED workflow

assumptionsandg committed:
Use replace instead of lineinfile

Alex-Welsh commented on .github/workflows/package-build-ofed.yml:
```diff
@@ -0,0 +1,246 @@
+      - name: Output image tag
+        id: image_tag
+        run: |
+          echo image_tag=$(grep stackhpc_rocky_$(sed s/-/_/ <(echo "9"))_overcloud_host_image_version: etc/kayobe/pulp-host-image-versions.yml | awk '{print $2}') >> $GITHUB_OUTPUT
```
The `sed` and `echo` magic isn't required here if the image is static. See [here](https://github.com/stackhpc/stackhpc-kayobe-config/blob/stackhpc/2024.1/.github/workflows/overcloud-host-image-build.yml#L52) for an example

holliefae replied:
Done 400f27f4

Alex-Welsh commented on etc/kayobe/pulp.yml:
```diff
@@ -412,6 +412,19 @@ stackhpc_pulp_distribution_rpm_production: >-
+stackhpc_pulp_doca_ofed_version: 24.07
+
+# DOCA OFED repositories
+stackhpc_pulp_repo_doca_ofed_rhel9:
```
Is there any plan for splitting this into dev/prod distributions?

Alex-Welsh replied:
Also, this content may deserve its own file. Not sure on that yet though

holliefae replied:
We're not making use of promote workflows yet. 

I think this can be accomplished in a similar way to the other RPM repositories with timestamping our builds though.

holliefae replied:
Release train support following this PR

Alex-Welsh commented on etc/kayobe/ansible/push-ofed.yml:
```diff
@@ -0,0 +1,54 @@
+          pulp \
+          --base-url '{{ stackhpc_release_pulp_url }}' \
+          --username '{{ stackhpc_release_pulp_username }}' \
+          --password '{{ stackhpc_release_pulp_password }}' \
```
Tad sketchy but since it's no_log it should be fine. I'm guessing there's no squeezer support for uploading a single package?

holliefae replied:
No I don't think so.

Alex-Welsh reviewed (commented):
Looks good as a first draft. We can merge soon if you want to start testing for real

assumptionsandg committed:
Remove sed magic

assumptionsandg committed:
Move OFED repositories to ofed.yml

assumptionsandg committed:
Rename build-ofed to build-ofed-rocky

assumptionsandg committed:
Add precheck for noexec

assumptionsandg committed:
Update workflow

assumptionsandg committed:
WIP: OFED documentation

assumptionsandg committed:
Fix no eol

assumptionsandg committed:
Add a release note

assumptionsandg committed:
Add to docs tree

Alex-Welsh commented on doc/source/contributor/ofed.rst:
There's room for improvement here but I don't want to hold up this PR any longer. At some point next week we should go through this together and make sure it says everything we want it to say

Alex-Welsh reviewed (approved):
Lets not hold this up any longer. There's plenty of improvements we could make but we shouldn't try to start with perfection.

Time to test in prod!
